### PR TITLE
feed `input_embeds` into `FlaxT5ForConditionalGeneration`

### PR DESCRIPTION
**Pull request by BigRedT**
This is a PR requested by @sanchit-gandhi in [https://github.com/huggingface/transformers/issues/18036#issuecomment-1214131955](https://github.com/huggingface/transformers/issues/18036#issuecomment-1214131955)

To summarize the issue - Flax encoder-decoder models are currently missing `input_embeds` argument, unlike the non-flax models. In this PR, I have added this argument for `FlaxT5ForConditionalGeneration` model and showed how this may be used for feeding features from other modalities such as vision into a language model such as `T5`.

Please run [examples/flax/vision-language/t5_for_vl.py](examples/flax/vision-language/t5_for_vl.py) for testing this feature.

Here's the output you should see:
```
--------------------------------------------------------------------------------
Model Input -> summarize: The US has "passed the peak" on new coronavirus cases, President Donald Trump said and predicted that some states would reopen this month. The US has over 637,000 confirmed Covid-19 cases and over 30,826 deaths, the highest for any country in the world. At the daily White House coronavirus briefing on Wednesday, Trump said new guidelines to reopen the country would be announced on Thursday after he speaks to governors.
--------------------------------------------------------------------------------
Summary from input_ids -> the country has over 637,000 confirmed cases and more than 30,826 deaths . the latest cases could be announced monday after speaking to governors .
--------------------------------------------------------------------------------
Summary from input_embeds -> the country has over 637,000 confirmed cases and more than 30,826 deaths . the latest cases could be announced monday after speaking to governors .
--------------------------------------------------------------------------------
Summary after concatenating random visual embeddings ->  the country has over 637,000 confirmed cases and more than 30,826 deaths . the latest cases could be announced monday after a tuesday vote .
```

### CODE REVIEW

**Comment by sanchit-gandhi** [comment]
Hey @BigRedT! Awesome that you've managed to get it working with so little code! And great to see that the model outputs are the same for `input_ids` and `input_embeds`. Before we can merge this we'll need to add some tests to make sure the functionality is as expected (which is should hopefully be given the toy example passes!). At the very least, we should add one test that mirrors the PyTorch test:
https://github.com/huggingface/transformers/blob/1ccd2515ed6d7da4ec46fe94aedbd8a86a2cde8e/tests/test_modeling_common.py#L2094
And one that verifies that the output logits for `input_ids` and `input_embeds` match. Do you want to have a go at this?

**Comment by sanchit-gandhi** [comment]
If you have any questions/issues, feel free to reach out to @patrickvonplaten or @patil-suraj. They will be more than happy to lend you a hand and provide a review on this PR! Otherwise I can take a look in a little over a weeks time! Thanks @BigRedT!

**Comment by BigRedT** [comment]
@sanchit-gandhi thanks for helping with this! Will take a look in the coming week.

**Comment by BigRedT** [comment]
Added `test_input_embeds()` to `FlaxT5ModelTest` in `test_modeling_flax_t5.py` as requested by @sanchit-gandhi 

This test currently checks to see if the generated sequences from `input_ids` match those from `input_embeds` (obtained by feeding `input_ids` through the `shared` embedding layer in the model)

@sanchit-gandhi wanted to see if the logits match too. @patrickvonplaten @patil-suraj what's the easiest way to compute logits. Also, could one of you review this PR. Thanks!

**Comment by BigRedT** [comment]
@sanchit-gandhi any updates?

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 1]
```diff
@@ -732,8 +732,13 @@ def __call__(
         return_dict: bool = True,
         deterministic: bool = True,
         init_cache: bool = False,
+        input_embeds=None
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:735]: Can we keep this arg name consistent with the PyTorch version?
https://github.com/huggingface/transformers/blob/c60dd98e87373e7f0f5af29f3d49411c2e81fb69/src/transformers/models/t5/modeling_t5.py#L898

```suggestion
        inputs_embeds=None
```

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 2]
```diff
@@ -732,8 +732,13 @@ def __call__(
...
     ):
-        hidden_states = self.embed_tokens(input_ids)
+        if input_embeds is not None:
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:737]: (nit) Could we write this `if`/`else` statement in the reverse order to which it currently is? Would improve the clarity!

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 3]
```diff
@@ -966,6 +971,7 @@ def __call__(
         train: bool = False,
         params: dict = None,
         dropout_rng: PRNGKey = None,
+        input_embeds: Optional[jnp.ndarray] = None
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:974]: Same again regarding the name

```suggestion
        inputs_embeds: Optional[jnp.ndarray] = None
```

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 4]
```diff
@@ -981,7 +987,10 @@ def __call__(
 
         # prepare encoder inputs
         if attention_mask is None:
-            attention_mask = jnp.ones_like(input_ids)
+            if input_ids is not None:
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:990]: Instead of a nested set of `if`/`else` statements, could we first establish the batch-size and seq-len from either the `input_ids` or `inputs_embeds`:
```python
if input_embeds is None:
    batch_size, sequence_length = input_ids.shape
else:
    batch_size, sequence_length = input_embeds.shape[:-1]
```
And then feed these into the attention mask:
```python
if attention_mask is None:
    attention_mask = jnp.ones(batch_size, sequence_length)
```
This sort of style helps a lot with readability! In general with Transformers, we advocate for longer, more readable code than concise, condensed code.

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 5]
```diff
@@ -990,9 +999,12 @@ def __call__(
         # Handle any PRNG if needed
         rngs = {"dropout": dropout_rng} if dropout_rng is not None else {}
 
+        if input_ids is not None:
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1002]: Will require the parallel statement for `inputs_embeds`:
```python
if inputs_embeds is not None:
    inputs_embeds = jnp.array(inputs_embeds, dtype="i4")
```

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 6]
```diff
@@ -1051,6 +1064,7 @@ def encode(
         train: bool = False,
         params: dict = None,
         dropout_rng: PRNGKey = None,
+        input_embeds: Optional[jnp.ndarray] = None
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1067]: Same again regarding the variable name!

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 7]
```diff
@@ -1074,7 +1088,11 @@ def encode(
         return_dict = return_dict if return_dict is not None else self.config.return_dict
 
         if attention_mask is None:
-            attention_mask = jnp.ones_like(input_ids)
+            if input_ids is None:
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1091]: Same again regarding the nested `if`/`else` loop!

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 8]
```diff
@@ -1074,7 +1088,11 @@ def encode(
...
+                attention_mask = jnp.ones(input_embeds.shape[:-1])
+            else:
+                attention_mask = jnp.ones_like(input_ids)
+                input_ids = jnp.array(input_ids, dtype="i4")
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1095]: Same again regarding needing the `jnp.array` conversion for `inputs_embeds`!

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 9]
```diff
@@ -1001,6 +1013,7 @@ def __call__(
             return_dict=return_dict,
             deterministic=not train,
             rngs=rngs,
+            input_embeds = input_embeds
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1016]: (nit) Could you move this variable position to the last position for the `__call__` method (i.e. after `deterministic`)?

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 10]
```diff
@@ -1087,14 +1105,15 @@ def _encoder_forward(module, input_ids, attention_mask, **kwargs):
...
-            input_ids=jnp.array(input_ids, dtype="i4"),
+            input_ids=input_ids,
             attention_mask=jnp.array(attention_mask, dtype="i4"),
             output_attentions=output_attentions,
             output_hidden_states=output_hidden_states,
             return_dict=return_dict,
             deterministic=not train,
             rngs=rngs,
             method=_encoder_forward,
+            input_embeds=input_embeds
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1116]: Same again regarding variable position!

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 11]
```diff
@@ -1480,6 +1502,9 @@ def setup(self):
             dtype=self.dtype,
         )
 
+    def get_shared_layer(self):
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1505]: Is this method strictly necessary? One can access the shared layers through `model.shared` (or something to that effect!) and the params through `model.params`

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 12]
```diff
@@ -1491,9 +1516,18 @@ def __call__(
         output_hidden_states=None,
         return_dict=None,
         deterministic: bool = True,
+        input_embeds=None,
     ):
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
+        if input_ids is not None:
+            B,L = input_ids.shape
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1524]: Could we make this more verbose? (`batch_size, sequence_length = ...`)

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 13]
```diff
@@ -1200,7 +1219,10 @@ def _decoder_forward(module, decoder_input_ids, decoder_attention_mask, **kwargs
 
         return outputs
 
+    def get_shared_layer(self):
```
Comment (sanchit-gandhi) [inline src/transformers/models/t5/modeling_flax_t5.py:1222]: Is this method strictly necessary?

File: src/transformers/generation_flax_utils.py [hunk 14]
```diff
@@ -279,7 +279,13 @@ def generate(
...
                 model_kwargs = self._prepare_encoder_decoder_kwargs_for_generation(input_ids, params, model_kwargs)
             # prepare decoder_input_ids for generation
-            input_ids = jnp.ones((input_ids.shape[0], 1), dtype="i4") * decoder_start_token_id
+            input_embeds = model_kwargs.get('input_embeds',None)
+            if input_embeds is None:
+                batch_size = input_ids.shape[0]
+            else:
+                batch_size = input_embeds.shape[0]
+            # Note: input_ids in the following line should really be decoder_input_ids
+            input_ids = jnp.ones((batch_size, 1), dtype="i4") * decoder_start_token_id
```
Comment (sanchit-gandhi) [inline src/transformers/generation_flax_utils.py:288]: Nice!

**Review by sanchit-gandhi — COMMENTED** [review]
Hey @BigRedT! Thanks for this PR and apologies for the delay in getting you a PR review.

Really like how you've managed to implement this change with very small code changes! That's right in the philosophy of Transformers! Before we can get this merged, it would be awesome just to clarify some of the new changes and make the code a bit more verbose. This helps immensely with readability and helps other's understand your cool new feature so that they too can use it themselves!

**Comment by sanchit-gandhi** [comment]
Hey @BigRedT, let me know if you want any further clarification for the comments. Happy to answer any questions! This PR looks pretty close to completion :)

**Comment by sanchit-gandhi** [comment]
Hey @BigRedT - do you want to see this PR to completion? We're pretty close now! 🤗
